src/libclipper/src/task_executor.cpp

```diff
@@ -59,21 +62,68 @@ void PredictionCache::put(const VersionedModelId &model,
                           const Output &output) {
   std::unique_lock<std::mutex> l(m_);
   auto key = hash(model, input->hash());
-  auto search = cache_.find(key);
-  if (search != cache_.end()) {
-    if (!search->second.completed_) {
+  auto search = entries_.find(key);
+  if (search != entries_.end()) {
+    CacheEntry &entry = search->second;
+    if (!entry.completed_) {
       // Complete the outstanding promises
-      for (auto &p : search->second.value_promises_) {
+      for (auto &p : entry.value_promises_) {
         p.set_value(output);
       }
-      search->second.completed_ = true;
-      search->second.value_ = output;
+      entry.completed_ = true;
+      entry.value_ = output;
+      size_bytes_ += output.y_hat_.size();
+      evict_entries(size_bytes_ - max_size_bytes_);
     }
   } else {
     CacheEntry new_entry;
     new_entry.value_ = output;
     new_entry.completed_ = true;
-    cache_.insert(std::make_pair(key, std::move(new_entry)));
+    insert_entry(key, new_entry);
+  }
+}
+
+void PredictionCache::insert_entry(const long key, CacheEntry &value) {
+  size_t entry_size_bytes = value.completed_ ? value.value_.y_hat_.size() : 0;
+  if (entry_size_bytes <= max_size_bytes_) {
+    evict_entries(size_bytes_ + entry_size_bytes - max_size_bytes_);
+    page_buffer_.insert(page_buffer_.begin() + page_buffer_index_, key);
```

Comment:
The `page_buffer_` is a vector of input hashes and acts as the clock in the cache eviction algorithm. In conjunction with the buffer (clock) itself, `page_buffer_index_` acts as the clock hand. To examine a page buffer entry with key `K`, we query the map `entries_` with `K` as the input hash key. We then determine whether or not the entry has all of its futures completed (via `entry.completed_`) and whether or not it was recently used (via the `entry.used_` boolean flag). If the entry is completed and was not recently used, we remove the key `K` from the page buffer, and the entry itself is removed from the `entries_` map.